### ISSUE TITLE
`bdep ci` seeing non-existing changes briefly

### ISSUE THREAD

**Issue by Klaim** [build2/build2#196]
I stumbled several times through years on this minor issue but it's so minor I keep forgetting to report it (and I'm not sure build2 is the source of the error or not):

```
Klaim@KLAIMSTATION2 MINGW64 /e/projects/build2-packaging/range-v3/range-v3 (master)
$ bdep ci
error: project directory has uncommitted changes
  info: run 'git status' for details
  info: use 'git stash' to temporarily hide the changes

Klaim@KLAIMSTATION2 MINGW64 /e/projects/build2-packaging/range-v3/range-v3 (master)
$ git status
On branch master
Your branch is up to date with 'origin/master'.

nothing to commit, working tree clean

Klaim@KLAIMSTATION2 MINGW64 /e/projects/build2-packaging/range-v3/range-v3 (master)
$ bdep ci
submitting:
  to:      https://ci.stage.build2.org
  in:      https://github.com/build2-packaging/range-v3.git#master@2c26749fa00d2df1086e4ea8fc7c89beddef9c85        
  package: range-v3
  version: 0.12.0
continue? [y/n] y
```

AAs you can see, `bdep ci` first detect changes, but `git status` say there is none, then `bdep ci` dont detect changes. I suspect that the execution of `git status` did change something that is what `bdep ci` relies on, but then it means `bdep ci` is probably not looking at the right data?

Unfortunately I don't know how to reproduce this for now, which is why I'm reporting this time I just stumbled upon it again.

As said this is a very minor issue, it is probably confusing for beginners though (and maybe erode trust in tools).

**Comment by boris-kolpackov** [comment]
Hm, interesting. Does this persist until you run `git status`? In other words, after hitting this error, if you keep re-running `bdep-ci`, do you keep getting the same error consistently or does it disappear after some time? If you haven't tried this, could you try next time you hit it?

**Comment by Klaim** [comment]
Of course. The git-bash version is the same as the git version:
```
$ git --version
git version 2.36.1.windows.1
```
(note that I installed git-bash with the option to use `cmd` as the base, not Cygwin)

Also while I saw this issue in the past and it migt not be important, right now I was using a weeks old `build2` stage version:
```
$ b --version
build2 0.15.0-a.0.3ec24f8b32f4
libbutl 0.15.0-a.0.c1e9c118b890
host x86_64-microsoft-win32-msvc14.2
```

**Comment by boris-kolpackov** [comment]
The `git` version in build2-baseutils for 0.15.0 is 2.33.1. So I think we will hold off on investigating this until we upgrade `git` to 2.36.1 (or later; this will happen at the beginning of the 0.16.0 development cycle). If the issue is still present, then we will investigate this further.

**Comment by boris-kolpackov** [comment]
@Klaim Would you be able to recheck if this still reproduces with the latest staged version of build2 (which comes with `git` `2.37.2.windows.2`)? Thanks!

**Comment by Klaim** [comment]
@boris-kolpackov I just upgraded to 
```
$ b --version
build2 0.16.0-a.0.456e2179dcb1
libbutl 0.16.0-a.0.bb5b2ed9b340
host x86_64-microsoft-win32-msvc14.3
$ bpkg --version
bpkg 0.16.0-a.0.545e427106ba
libbpkg 0.16.0-a.0.7c37a8d9a6c6
libbutl 0.16.0-a.0.bb5b2ed9b340
$ bdep --version
bdep 0.16.0-a.0.0e9a9ec0c6f5
libbpkg 0.16.0-a.0.7c37a8d9a6c6
libbutl 0.16.0-a.0.bb5b2ed9b340
```

I can reproduce the issue as described above. The only change is that the first time I run `bdep ci` seems to take several seconds and the cpu charge rises suddenly (I can hear the fans which usually are quiet). It might jsut be teh antivirus checking teh exes though.
Other than that, no change, I get 
```
$ bdep ci
error: project directory has uncommitted changes
  info: run 'git status' for details
  info: use 'git stash' to temporarily hide the changes
```
as before (I tried 6 times, same output) and running `git status` fixes `bdep ci` as before.

I did not upgrade my system git version which is `2.36.1.windows.1` assuming it wouldnt impact the issue. I upgraded to `2.41.0.windows.1` after that first test repro and re-tried just to be sure and as expected it had no impact.

Looks like the issue isnt from the embedded git version?

**Comment by boris-kolpackov** [comment]
Thanks for testing! It's unclear what's going on, feels like one of those "Windows happens" situations. Hopefully will find time to investigate it before the release.

**Comment by Klaim** [comment]
Yeah at least there is a simple workaround, this is not urgent to fix clearly.

**Comment by boris-kolpackov** [comment]
For completeness, I've tried your instructions on Linux and naturally everything works as expected. So this seems to be a Windows/git-specific issues. I am not sure when we will have time to look into this (there is just too much to do) so if someone who cares about Windows can investigate this further, that would be great.

**Comment by kamrann** [comment]
I've also run into this sporadically, using Powershell in my case.

What I'm hitting far more frequently though, is the fact that `bdep` (both `ci` and `release`) seems to report uncommitted changes when there are untracked files but no changes to anything in the index. This appears to always be the case, as opposed to the sporadic behaviour detailed above. Is it intentional?

In my experience, `git` commands (`rebase` for example) will generally require there to be no index changes, but untracked files do not affect things. Also, the message says "info: use 'git stash' to temporarily hide the changes", which could cause confusion in these cases - admittedly `git stash -a` will work, but for above stated reasons I've never had to use that before and didn't even know it existed until now.

**Comment by boris-kolpackov** [comment]
Yes, this is intentional. The reason we warn is that what you `ci`/`release` does not match the state of your source tree. The untracked files could be related or not to the command in question, we cannot guess that. But I believe it's better to ask the user to sort it out than potentially produce an incorrect result.

**Comment by kamrann** [comment]
Ok understood. I'd suggest maybe adding `-u` to the suggested `stash` command, and/or adding a `--force` or similar option to allow for expressly ignoring this.
